smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/IncomingKafkaRecordBatch.java

```diff
@@ -0,0 +1,104 @@
+
+public class IncomingKafkaRecordBatch<K, T> implements KafkaRecordBatch<K, T> {
+
+    private final Metadata metadata;
+    private final List<KafkaRecord<K, T>> incomingRecords;
+    private final Map<TopicPartition, KafkaRecord<K, T>> latestOffsetRecords;
+
+    public IncomingKafkaRecordBatch(ConsumerRecords<K, T> records, KafkaCommitHandler commitHandler,
+            KafkaFailureHandler onNack, boolean cloudEventEnabled, boolean tracingEnabled) {
+        List<IncomingKafkaRecord<K, T>> incomingRecords = new ArrayList<>();
+        Map<TopicPartition, IncomingKafkaRecord<K, T>> latestOffsetRecords = new HashMap<>();
+        for (TopicPartition partition : records.partitions()) {
+            IncomingKafkaRecord<K, T> latestOffsetRecord = null;
+            for (ConsumerRecord<K, T> record : records.records(partition)) {
+                IncomingKafkaRecord<K, T> rec = new IncomingKafkaRecord<>(record, commitHandler, onNack,
+                        cloudEventEnabled, tracingEnabled);
+                incomingRecords.add(rec);
+                latestOffsetRecords.put(partition, rec);
+                latestOffsetRecord = rec;
+            }
+            if (latestOffsetRecord != null) {
+                commitHandler.received(latestOffsetRecord);
+            }
+        }
+        this.incomingRecords = Collections.unmodifiableList(incomingRecords);
+        this.latestOffsetRecords = Collections.unmodifiableMap(latestOffsetRecords);
+        this.metadata = Metadata.of(new IncomingKafkaRecordBatchMetadata<>(records));
+    }
+
+    @Override
+    public List<T> getPayload() {
+        return this.incomingRecords.stream().map(KafkaRecord::getPayload).collect(Collectors.toList());
+    }
+
+    @Override
+    public List<KafkaRecord<K, T>> getRecords() {
+        return this.incomingRecords;
+    }
+
+    @Override
+    public Iterator<KafkaRecord<K, T>> iterator() {
+        return this.getRecords().iterator();
+    }
+
+    @Override
+    public Map<TopicPartition, KafkaRecord<K, T>> getLatestOffsetRecords() {
+        return this.latestOffsetRecords;
+    }
+
+    @Override
+    public Metadata getMetadata() {
+        return metadata;
+    }
+
+    @Override
+    public Supplier<CompletionStage<Void>> getAck() {
+        return this::ack;
+    }
+
+    @Override
+    public Function<Throwable, CompletionStage<Void>> getNack() {
+        return this::nack;
+    }
+
+    @Override
+    public CompletionStage<Void> ack() {
+        return Multi.createBy().concatenating().collectFailures()
+                .streams(this.latestOffsetRecords.values().stream()
+                        .map(record -> Multi.createFrom().completionStage(record::ack))
+                        .collect(Collectors.toList()))
+                .toUni().convert().toCompletionStage();
+    }
+
+    @Override
+    public CompletionStage<Void> nack(Throwable reason, Metadata metadata) {
+        return Multi.createBy().concatenating().collectFailures()
```

Comment:
should we concatenate or merge?

Reply:
I went for concatenate to preserve the ordering.